src/data_race.rs

```diff
@@ -0,0 +1,1189 @@
+            Ok(())
+        }else{
+            Ok(())
+        }
+    }
+
+    /// Detect data-races for an unsychronized write operation, will not perform
+    ///  data-race threads if `multi-threaded` is false, either due to no threads
+    ///  being created or if it is temporarily disabled during a racy read or write
+    ///  operation
+    pub fn write<'tcx>(&mut self, pointer: Pointer<Tag>, len: Size) -> InterpResult<'tcx> {
+        self.unique_access(pointer, len, "Write")
+    }
+    /// Detect data-races for an unsychronized deallocate operation, will not perform
+    ///  data-race threads if `multi-threaded` is false, either due to no threads
+    ///  being created or if it is temporarily disabled during a racy read or write
+    ///  operation
+    pub fn deallocate<'tcx>(&mut self, pointer: Pointer<Tag>, len: Size) -> InterpResult<'tcx> {
+        self.unique_access(pointer, len, "Deallocate")
+    }
+}
+
+/// The current set of vector clocks describing the state
+///  of a thread, contains the happens-before clock and
+///  additional metadata to model atomic fence operations
+#[derive(Clone, Default, Debug)]
+struct ThreadClockSet {
+
+    /// The increasing clock representing timestamps
+    ///  that happen-before this thread.
+    clock: VClock,
+
+    /// The set of timestamps that will happen-before this
+    ///  thread once it performs an acquire fence
+    fence_acquire: VClock,
+
+    /// The last timesamp of happens-before relations that
+    ///  have been released by this thread by a fence
+    fence_release: VClock,
+}
+
+impl ThreadClockSet {
+
+    /// Apply the effects of a release fence to this
+    ///  set of thread vector clocks
+    #[inline]
+    fn apply_release_fence(&mut self) {
+        self.fence_release.clone_from(&self.clock);
+    }
+
+    /// Apply the effects of a acquire fence to this
+    ///  set of thread vector clocks
+    #[inline]
+    fn apply_acquire_fence(&mut self) {
+        self.clock.join(&self.fence_acquire);
+    }
+
+    /// Increment the happens-before clock at a
+    ///  known index
+    #[inline]
+    fn increment_clock(&mut self, index: VectorIdx) {
+        self.clock.increment_index(index);
+    }
+
+    /// Join the happens-before clock with that of
+    ///  another thread, used to model thread join
+    ///  operations
+    fn join_with(&mut self, other: &ThreadClockSet) {
+        self.clock.join(&other.clock);
+    }
+}
+
+/// Extra metadata associated with a thread
+#[derive(Debug, Clone, Default)]
+struct ThreadExtraState {
+
+    /// The current vector index in use by the
+    ///  thread currently, this is set to None
+    ///  after the vector index has been re-used
+    ///  and hence the value will never need to be
+    ///  read during data-race reporting
+    vector_index: Option<VectorIdx>,
+
+    /// The name of the thread, updated for better
+    ///  diagnostics when reporting detected data
+    ///  races
+    thread_name: Option<Box<str>>,
```

Comment:
Isn't this information already present in the concurrency subsystem? Why copy it here?